File: mat_analysis.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 19 15:42:30 2018

Performs analysis of distance and point clustering (Sadahiro & Takami, 2001) for each shape.

Distance data will be saved in the directory "distance_analysis" with a subdirectory for each shape mask. 
Each file contains data to create a probability distribution defined by the generated uniform sets.

The MATLAB script StatAnalysis.m will use these to calculate the significance of the observed data. 

@author: Jamie Dunkle
"""

import cv2
import os, errno
import numpy as np
import scipy.io as sio
import distances as dist
from timeit import default_timer as timer


################## Globals - CHANGE THESE TO RUN ON SPECIFIC SUBJECTS AND SHAPE SETS
analysis_conds = ["bounding_circle","in_shape","touchpoint_hull"]
img_names = ["solo3","solo5","solo6","solo7","solo9","solo10","solo11","solo12",
             "blake_01","blake_04","blake_06","blake_07","blake_08","blake_09","blake_10","blake_11","blake_12"]
patient = "MC"   
img_path = "./Shapes/"         # path containing shape images



################# Function Definitions #########################################################################
def getVarMean(data) :
    var = np.sum(np.power(data, 2))/(np.size(data)-1)
    rmse = np.sqrt(var)
    amd = np.mean(data)
    return [var, rmse, amd]


def getVarMean2D(data) :
    var = np.sum(np.power(data,2),axis=1) / (data.shape[1] - 1)
    rmse = np.sqrt(var)
    amd = np.mean(data, axis=1)
    return [var, rmse, amd]


# # cumulative distribution function for "reference object" defined by ro_pts
# # returns ratio of points located within a given distance of the reference object
# def cdf(points_in, ro_pts, max_r):
#     # points2points_cmin crashes for points > 250, so fall back to numpy function
#     if points_in.shape[0] > 250 : 
#         ft_dists = dist.points2points_np(points_in,ro_pts)
#     else :
#         ft_dists = dist.points2points_c(points_in,ro_pts)
#     points_in_region = np.empty((max_r+1,1), dtype=int)
#     for i in range(max_r+1) : 
#         points_in_region[i] = (ft_dists <= i).sum()
#     return np.true_divide(points_in_region, points_in.shape[0])


def get_dvals(ro_pts, max_r, generated_unif, observed, generated) :
    # calculate d-values for main shape
    ds_gu = dist.cdf(generated_unif, ro_pts, max_r)
    # calculate d-values for observed data
    ds_o = dist.cdf(observed, ro_pts, max_r)
    diff_o = ds_o - ds_gu
    d_o_plus = np.max(diff_o)
    d_o_plus_r = np.argmax(diff_o)
    d_o_minus = np.min(diff_o)
    d_o_minus_r = np.argmin(diff_o)
    # calculate d-values for 100k simulated data sets --> CDF
    ds_g = np.zeros((generated.shape[0],max_r), dtype=np.float32)
    diff_g = np.zeros((generated.shape[0],max_r), dtype=np.float32)
    for i in range(generated.shape[0]) :
        ds_g[i,:] = dist.cdf(generated[i], ro_pts, max_r)
        diff_g[i,:] = ds_g[i,:] - ds_gu
    d_g_plus = np.max(diff_g, axis=1)
    d_g_minus = np.min(diff_g, axis=1)
    return d_o_plus, d_o_plus_r, d_o_minus, d_o_minus_r, d_g_plus, d_g_minus


def matAnalysis(img_file, img_path, img_mat, mat_path, obs_mat, obs_path, gen_mat, gen_path, out_path):    

    # Read in the image 
    img = cv2.imread(os.path.join(img_path, img_file) ,cv2.IMREAD_UNCHANGED)
    img[(img[:,:,3]==0),0:3] = 0
    
    try:
        shape_analysis = sio.loadmat(os.path.join(mat_path, img_mat))
    except (TypeError, IOError) :
        print "Error loading: {0} -- skipping {1}...".format(img_mat, img_name)
        return
 
    ma_points = np.ascontiguousarray(shape_analysis['ma_points'].astype(np.float32)) # (x,y)
    edge_points = np.ascontiguousarray(shape_analysis['edge_points'].astype(np.float32)) # (x,y)
    centroid = np.ascontiguousarray(shape_analysis['centroid'].astype(np.float32)) # (x,y)
    
    #print "Calculating Observed"
    try:
        observed_mat = sio.loadmat(os.path.join(obs_path, obs_mat))
    except (TypeError, IOError) :
        print "Error loading: {0} -- skipping {1}...".format(obs_mat, img_name)
        return
    observed = np.ascontiguousarray(observed_mat['img_dataset'].astype(np.float32))
    observed[:,1] = img.shape[0]-observed[:,1] # opencv coordinates use inverted y-axis
    
    if cond == "in_shape" :
        observed_inshape = [] 
        for i in range(observed.shape[0]) :
            if cv2.pointPolygonTest(edge_points, tuple(observed[i]), measureDist=False) == 1 :
                observed_inshape.append(observed[i])
        observed = np.array(observed_inshape)
    
    start = timer()
    observed_ma_dists = dist.points2points_c(observed,ma_points)
    observed_edge_dists = dist.points2points_c(observed,edge_points)
    observed_centroid_dists = dist.points2points_c(observed,centroid)

     # Save data
    observed_medaxis_data = getVarMean(observed_ma_dists)
    observed_edge_data = getVarMean(observed_edge_dists)
    observed_centroid_data = getVarMean(observed_centroid_dists)
   
#    #print "Calculating Generated"
    generated_mat = sio.loadmat(os.path.join(gen_path, gen_mat))
    generated = np.ascontiguousarray(generated_mat['gen_datasets'].astype(np.float32))
    
    generated_ma_dists = np.empty((generated.shape[0],generated.shape[1]))
    generated_edge_dists = np.empty((generated.shape[0],generated.shape[1]))
    generated_centroid_dists = np.empty((generated.shape[0],generated.shape[1]))
    
    for i, point_set in enumerate(generated) :
        generated_ma_dists[i] = dist.points2points_c(point_set,ma_points)
        generated_edge_dists[i] = dist.points2points_c(point_set,edge_points)
        generated_centroid_dists[i] = dist.points2points_c(point_set,centroid)

    #Save data
    generated_medaxis_data = getVarMean2D(generated_ma_dists)
    generated_edge_data = getVarMean2D(generated_edge_dists)
    generated_centroid_data = getVarMean2D(generated_centroid_dists)
    generated_medaxis_data = [list(datum) for datum in generated_medaxis_data]
    generated_edge_data = [list(datum) for datum in generated_edge_data]
    generated_centroid_data = [list(datum) for datum in generated_centroid_data]
    print "var-mean data:", timer() - start, "s"
    
    sio.savemat(os.path.join(out_path,img_name+'_analysis_var_mean.mat'),
                {'observed_medaxis_data':observed_medaxis_data, 
                 'observed_edge_data':observed_edge_data, 
                 'observed_centroid_data':observed_centroid_data,
                 'generated_medaxis_data':generated_medaxis_data, 
                 'generated_edge_data':generated_edge_data,
                 'generated_centroid_data':generated_centroid_data})
    
    img_bin = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
    img_bin[(img_bin!=0)] = 1 # convert to logical array (white = 1)
    white_idx = np.where(img_bin==1)    
    shape_points = np.transpose(np.array(white_idx))
    shape_points = np.ascontiguousarray(np.flip(shape_points, axis=1).astype(np.float32))
    
    max_r = np.max(img.shape[:2])/2
    
    start = timer()
    dm_o_plus, dm_o_plus_r, dm_o_minus, dm_o_minus_r, dm_g_plus, dm_g_minus = get_dvals(ma_points, max_r, shape_points, observed, generated)
    print "spat medial axis data:", timer() - start, "s"
    
    start = timer()
    de_o_plus, de_o_plus_r, de_o_minus, de_o_minus_r, de_g_plus, de_g_minus = get_dvals(edge_points, max_r, shape_points, observed, generated)
    print "spat edge data:", timer() - start, "s"
    
    start = timer()
    dc_o_plus, dc_o_plus_r, dc_o_minus, dc_o_minus_r, dc_g_plus, dc_g_minus = get_dvals(centroid, max_r, shape_points, observed, generated)
    print "spat centroid data:", timer()-start, "s\n"
    
    sio.savemat(os.path.join(out_path,img_name+'_analysis_spat.mat'),
                {'observed_medaxis_dplus':[dm_o_plus,dm_o_plus_r],
                 'observed_medaxis_dminus':[dm_o_minus,dm_o_minus_r],
                 'observed_edge_dplus':[de_o_plus,de_o_plus_r],
                 'observed_edge_dminus':[de_o_minus,de_o_minus_r],
                 'observed_centroid_dplus':[dc_o_plus,dc_o_plus_r],
                 'observed_centroid_dminus':[dc_o_minus,dc_o_minus_r],
                 'generated_medaxis_dplus':dm_g_plus,
                 'generated_medaxis_dminus':dm_g_minus,
                 'generated_edge_dplus':de_g_plus,
                 'generated_edge_dminus':de_g_minus,
                 'generated_centroid_dplus':dc_g_plus,
                 'generated_centroid_dminus':dc_g_minus})
    
    

if __name__ == '__main__':
    
    mat_path = patient+"/shape_analysis/"          # path containing medial axis mat files
    obs_path = patient+"/aggregated_observations/" # path containing observed data
    
    for cond in analysis_conds :
        print "Condition:", cond
        
        gen_path = patient+"/generated_uniform_data/"+cond+"/"  # path containing generated uniform data
        
        out_path = patient+"/distance_analysis/"+cond+"/"       # output path
        try:
            os.makedirs(out_path)
        except OSError as e:
            if e.errno != errno.EEXIST:
                raise

        for img_name in img_names :
            print 'Starting', img_name
            img_file = img_name + ".png"
            img_mat = img_name + "_shape_analysis.mat"
            obs_mat = img_name + "_Patient_"+patient+"_aggregated_observations.mat"
            gen_mat = img_name + "_Patient_"+patient+"_generated_uniform_sets.mat"
            matAnalysis(img_file, img_path, img_mat, mat_path, obs_mat, obs_path, gen_mat, gen_path, out_path)

File: mat_analysis.m
% import cv2
% import numpy as np
% import scipy.io as sio
% import vectorization as vect
% 
% import assist

global nextCirc
nextCirc = [ [-1,-1]; [0,1]; [1,-1]; [1,0];...
             [1,1];   [0,1]; [-1,1]; [-1,0]; ];

global img
global img_size

% Define colors for drawing.
delaunay_color = [255,255,255];
points_color = [0, 0, 255];
    
img_names = ['blake_01';'blake_04';'blake_06';'blake_07';'blake_08';'blake_09';'blake_10';'blake_11';'blake_12']; %,"solo10","solo11","solo3","solo6","solo7","solo9"]
divvy = [18,22,11,18,10,16,21,21,19]; %,10,19,5,1,9,15]
patient = 'DF';
%shape_img = strcat('Blake/',strcat(img_names,'.png'));
%generated_files = strcat(['./' patient '/generated_results/1000/'],strcat(img_names,['_Patient_' patient '_generated_results.mat']));
%observed_files = strcat(['./' patient '/aggregated_observations/'],strcat(img_names,['_Patient_' patient '_aggregated_observations.mat']));
out_path = [patient '/test_mat/'];
%print observed_files
                
%super_matrix = zip(img_names,divvy,shape_img,generated_files,observed_files);
                
aggregate_ma = [];
aggregate_c = [];
                
%global_start = timer()
for f=1:length(img_names)
    
    %start_row = timer()
    
    % Read in the image.
    [img,map] = imread(['Blake/' img_names(f,:) '.png']);
    img = ind2rgb(img, map);
    img_orig = img; % Keep a copy around

    % Rectangle to be used with Subdiv2D
    img_size = size(img);
    rect = [0, 0, img_size(2), img_size(1)];

    % Create an instance of Subdiv2D
    %subdiv = cv2.Subdiv2D(rect);

    % Create an array of points.
    point = [round(img_size(1)/3),1];
    points = [];
    black = [round(img_size(1)/3),0];

    whites = [0,0];
    white_count = 0;     % make this a list of shapes - white counts
    for y=1:img_size(1)     % img_size(1) is number of rows
        for x=1:img_size(2) % img_size(2) is number of columns
            if isWhite([y,x])
                whites = whites + [y,x];
                white_count = white_count + 1;
            end
        end
    end
    centroid = [whites/white_count, 0];

    while 1 % go until you get to a white pixel
        if isWhite(point)
            points = [points; point];
            break
        else
            black = point;
            point(2) = point(2) + 1;
        end
    end

    while 1 % find all
        [black, nextPoint] = nextPixel(black, point, points);
        if all(nextPoint~=[-1,-1]) && ~isWhite(black)
            point = nextPoint;
            points = [points; point];
        else
            break
        end
    end
    
    count = 0;
    temp_points = [];
    for p=1:divvy(f):length(points)
        temp_points = [temp_points; points(p,:)];
    end
    points = temp_points;
    
    % Insert points into subdiv
    myCnt = [];
    for p=1:length(points)
        myCnt = [myCnt; points(p,:)];
        %subdiv.insert(points(p,:));
    end
    
    % Draw delaunay triangles
    %draw_delaunay( img, subdiv, delaunay_color );
    % Draw points
%     for p=1:length(points)
%         draw_point(img, points(p,:), points_color);
%     end

    % Allocate space for Voronoi Diagram
    img_voronoi = zeros(size(img));

    % Draw Voronoi diagram
    ma_lines = draw_voronoi(img_voronoi,subdiv,myCnt);
    % sio.savemat('./Patient_'+patient+'/ma_c_values/' + row[0] + '_ma.mat',{'ma_lines':ma_lines});
    % sio.savemat('./Patient_'+patient+'/ma_c_values/' + row[0] + '_c.mat',{'centroid':centroid});

    %print "Calculating Observed"
    % let's start computing distances!
    % observed first. Faster.
    %observed_mat = sio.loadmat(row(4));
    %observed = observed_mat['img_dataset']

    %start_time = timer()
    observed_min_dists = vect.dist_pts2lines(observed,ma_lines);
    observed_centroid_dists = vect.dist_pts2cent(observed,centroid);
    %print "Observed: ",(timer() - start_time)*1000, 'ms'

    observed_medaxis_var = sum(power(observed_min_dists, 2))/(size(observed_min_dists)-1);
    observed_medaxis_amd = mean(observed_min_dists);
    observed_medaxis_data = [observed_medaxis_var, observed_medaxis_amd];
%     sio.savemat(out_path+row(0)+'_medaxis_new.mat',{'observed_medaxis_data':observed_medaxis_data});

    observed_centroid_var = sum(power(observed_centroid_dists, 2))/(size(observed_centroid_dists)-1);
    observed_centroid_amd = mean(observed_centroid_dists);
    observed_centroid_data = [observed_centroid_var, observed_centroid_amd];
%     sio.savemat(out_path+row(0)+'_centroid_new.mat',{'observed_centroid_data':observed_centroid_data});

    %print "Calculating Generated"
    %generated_mat = sio.loadmat(row(3));
    %generated = generated_mat['img_datasets'];
    generated_centroid_data = [];
    generated_medaxis_data = [];

    generated_min_dists = zeros(generated.shape(0),generated.shape(1));
    generated_centroid_dists = zeros(generated.shape(0),generated.shape(1));
    %start_time = timer()
    for i=1:length(generated)%, point_set in enumerate(generated)
        generated_min_dists(i) = vect.dist_pts2lines(point_set,ma_lines);
        generated_centroid_dists(i) = vect.dist_pts2cent(point_set,centroid);
        %print "Generated: ",(timer() - start_time), 's'
    end
    % medial axis variables
    generated_medaxis_var = sum(np.power(generated_min_dists,2)) / (generated_min_dists.shape(1) - 1);
    generated_medaxis_amd = mean(generated_min_dists);%, axis=1)
    generated_medaxis_data = [generated_medaxis_var, generated_medaxis_amd];

    generated_centroid_var = sum(np.power(generated_centroid_dists,2)) / (generated_centroid_dists.shape(1) - 1);
    generated_centroid_amd = mean(generated_centroid_dists);
    generated_centroid_data = [generated_centroid_var, generated_centroid_amd];
end
    % print "saving file"
%     generated_medaxis_data = [list(datum) for datum in generated_medaxis_data]
%     generated_centroid_data = [list(datum) for datum in generated_centroid_data]
% 
%     sio.savemat(out_path+row[0]+'_generated_ma_new.mat',{'generated_medaxis_data':generated_medaxis_data})
%     sio.savemat(out_path+row[0]+'_generated_c_new.mat',{'generated_centroid_data':generated_centroid_data})

                
function w = isWhite(point)
    global img
    global img_size
    w = 0;
    if any(point > img_size(1:2)) || any(point <= 0)
        return
    elseif all(squeeze(img(point(1),point(2),:)))
        w = 1;
    end
end                                                                                
                                                                                
% finds the first white point that is not in points
function [black,newPoint] = nextPixel(black, point, points)
    global nextCirc
    for w=1:8
        pt_offset = point + nextCirc(w,:);
        % if offset point is white and not already in points
        if isWhite(pt_offset) && ~ismember(2,sum(points==pt_offset,2))
            newPoint = pt_offset;    
            return
        else
            black = pt_offset;
        end
    end
    newPoint = [-1,-1];
    return
end 

% Check if a point is inside a rectangle
function contains = rect_contains(rect, point)
    contains = 1;
    if any(point<rect(1:2)) || any(point>rect(3:4))
        contains = 0;
    end
end

% Draw a point
function draw_point(img, p, color )
    cv2.circle( img, p, 2, color, cv2.FILLED, cv2.LINE_AA, 0 );
end
 
%  
% % Draw delaunay triangles
% def draw_delaunay(img, subdiv, delaunay_color ) :
%  
%     triangleList = subdiv.getTriangleList();
%     img_size = img.shape
%     r = (0, 0, img_size[1], img_size[0])
%  
%     for t in triangleList :
%          
%         pt1 = (t[0], t[1])
%         pt2 = (t[2], t[3])
%         pt3 = (t[4], t[5])
%          
%         if rect_contains(r, pt1) and rect_contains(r, pt2) and rect_contains(r, pt3) :
%          
%             cv2.line(img, pt1, pt2, delaunay_color, 1, cv2.LINE_AA, 0)
%             cv2.line(img, pt2, pt3, delaunay_color, 1, cv2.LINE_AA, 0)
%             cv2.line(img, pt3, pt1, delaunay_color, 1, cv2.LINE_AA, 0)
%  
%  
% % Draw voronoi diagram
% def draw_voronoi(img, subdiv, myCnt) :
%     img_size = img.shape
%     rect = np.array([[0,0], [0,img_size[0]], [img_size[1],img_size[0]], [img_size[1],0]])
%     cv2.fillConvexPoly(img, rect, (255,255,255), cv2.LINE_AA, 0)
% 
%     [facets, centers] = subdiv.getVoronoiFacetList([])
%     lines = []
%     
%     for i in xrange(0,len(facets)) :
%         for f in range(0,len(facets[i])) :
%             if cv2.pointPolygonTest(np.array(myCnt), tuple(facets[i][f]), False) > 0 && cv2.pointPolygonTest(np.array(myCnt), tuple(facets[i][(f+1)%len(facets[i])]), False) > 0 :
%                 lines.append([facets[i][f], facets[i][(f+1);%len(facets[i])]])
%             end
%         cv2.circle(img, (centers[i][0], centers[i][1]), 3, (0, 0, 0), cv2.FILLED, cv2.LINE_AA, 0);
%         end
%     end
%         
%     for l in lines:
%         cv2.line(img, tuple(l[0]), tuple(l[1]), (0,0,0), 1, cv2.LINE_AA)
%         % print l[0][0], l[0][1], l[1][0], l[1][1]
%     return lines
